File: hackerrank/problem solving/algorithms/2 - implementation/46 - beautiful triplets/beautiful_triplets_procedural.rb
# Source: https://www.hackerrank.com/challenges/beautiful-triplets/problem?isFullScreen=true

def main
    size, beautiful_difference = read_an_array
    array = read_an_array

    n_beautiful_triplets = find_beautiful_triplets(array, beautiful_difference)
    puts n_beautiful_triplets
end

    def read_an_array
        array = gets.split.map(&:to_i)
    end

    def find_beautiful_triplets(array, beautiful_difference)
        n_beautiful_triplets = 0

        if array.size > 2
            for i in 0...(array.size-2)
                for j in (i+1)...(array.size-1)
                    first_difference = array[j] - array[i]

                    if first_difference < beautiful_difference
                        next
                    elsif first_difference > beautiful_difference
                        break
                    end
                    
                    for k in (j+1)...array.size
                        second_difference = array[k] - array[j]

                        if second_difference == beautiful_difference
                            n_beautiful_triplets += 1
                        elsif second_difference > beautiful_difference
                            break
                        end
                    end
                end
            end
        end

        n_beautiful_triplets
    end

main

File: hackerrank/problem solving/algorithms/2 - implementation/46 - beautiful triplets/beautiful_triplets_procedural.js
// Source: https://www.hackerrank.com/challenges/beautiful-triplets/problem?isFullScreen=true

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');
    main();
});

function readLine() {
    return inputString[currentLine++];
}


function main() {
    const [SIZE, BEAUTIFUL_DIFFERENCE] = readAnArray();
    const ARRAY = readAnArray();

    const N_BEAUTIFUL_TRIPLETS = findBeautifulTriplets(ARRAY, BEAUTIFUL_DIFFERENCE);
    console.log(N_BEAUTIFUL_TRIPLETS);
}

    function readAnArray() {
        const ARRAY = readLine().split(" ").map(Number);
        return ARRAY;
    }

    function findBeautifulTriplets(array, beautifulDifference) {
        let nBeautifulTriplets = 0;

        if (array.length > 2)
            for (let i = 0; i < array.length-2; i++)
                for (let j = i + 1; j < array.length-1; j++) {
                    const FIRST_DIFFERENCE = array[j] - array[i];

                    if (FIRST_DIFFERENCE < beautifulDifference)
                        continue;
                    else if (FIRST_DIFFERENCE > beautifulDifference)
                        break;

                    for (let k = j + 1; k < array.length; k++) {
                        const SECOND_DIFFERENCE = array[k] - array[j];

                        if (SECOND_DIFFERENCE == beautifulDifference)
                            nBeautifulTriplets++;
                        else if (SECOND_DIFFERENCE > beautifulDifference)
                            break;
                    }
                }

        return nBeautifulTriplets;
    }
